Merge core libraries.

grokys opened the pull request:
## What does the pull request do?

We have too much separation of concerns in the Avalonia assembly layout, which has caused (and is causing) various problems with developing Avalonia. Some examples:

- Hard to write features which animate classes in the `Avalonia.Visuals` assembly as that assembly was lower than `Avalonia.Animation`
- Impossible to unify `VisualChildren` and `LogicalChildren` into a single collection as they're defined in different assemblies

This PR merges all assemblies below (but not including) `Avalonia.Controls` into `Avalonia.Base`.

Open questions:

- Should we also put Avalonia.Markup in there?
- Should we also put the base control classes (`Control`, `TemplatedControl`) in there (moving the actual controls to e.g. `Avalonia.Controls.Standard` maybe?)

## Breaking changes

Pretty big ABI breaking change: a lot of classes have moved and assemblies have been deleted.

grokys committed:
Merge core libraries.

Everything below `Avalonia.Controls` into `Avalonia.Base`.

robloo commented:
My two cents:

> Should we also put Avalonia.Markup in there?

Markup makes sense to keep separate IMO. It's entirely possible within 10 years this layer needs to change independently anyway.

> Should we also put the base control classes (Control, TemplatedControl) in there (moving the actual controls to e.g. Avalonia.Controls.Standard maybe?)

It seems to make sense to move base classes into the unified base project. Then these base classes would be available from all 'controls' projects. However, I wouldn't then call the controls project `Avalonia.Controls.Standard` it should remain `Avalonia.Controls` as the standard part is strongly implied. Any additional controls packages would then be `Avalonia.Conrols.WinUI` etc.

maxkatz6 commented:
Another questions, do we have plans to change nuget packages? Right now we have couple of dlls merged into one "Avalonia" package. Do we want to drop that merging tool and publish simplified nuget packages?

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

jmacato committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Move new files to correct place.

grokys committed:
Removed unused dirs/projects,

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Removed outdated references from theme assemblies.

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Merge unit tests to match new assembly layout.

grokys committed:
Fixup test namespaces.

grokys committed:
Make directory match namespace.

grokys committed:
Move files to match namespace.

grokys committed:
Move files to match namespace.

grokys committed:
Fix up incorrect namespace.

`Avalonia.Visuals.Media.Imaging` -> `Avalonia.Media.Imaging`.

grokys committed:
Fix resource URL.

grokys committed:
Removed outdated dependencies.

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Added missing project reference.

rstm-sf commented:
Hi, what is the status of this PR and in what order will it be merged?

maxkatz6 commented:
@rstm-sf it was planned to be merged when we will be ready to with doing 0.11.0 branch with breaking changes instead of backporting everything to 0.10.x.

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Update test namespaces.

jmacato committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Fix merge error.

robloo commented:
Will this make Avalonia.Base internal members visible to Avalonia.Controls? Access to things such as KnownColors isn't possible otherwise. Internals need to be made available IMO. I already have found a few places it's needed/useful and haven't done much in the code base.

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys committed:
Fix merge errors.

robloo commented:
I see lots of issues potential issues with a massive commit like this.

First, is there any way you can do it project-by-project first rather than all at once?

Secondly, this change should be done immediately or it's just going to get worse. I would also notify all developers of a hard-stop in submitting PRs until this is done.

Third, all existing PRs are going to need to be re-done. (This is a good reason why the PRs list shouldn't be allowed to get so long)

grokys committed:
Merge branch 'master' into refactor/merge-core-assemblies

grokys commented:
Ok, this should be ready to go. Will be merged shortly.

grokys committed:
Fix bad merge in WindowsInteropTest.csproj.

grokys commented:
> I see lots of potential issues with a massive commit like this.

Definitely, which is why we've been holding off - it's potentially going to make backporting fixes a lot more difficult.

> First, is there any way you can do it project-by-project first rather than all at once?

I'm not sure that would make it any less painful. The plan is to merge this as a squashed commit so that the painful changes are all in one place.

> Secondly, this change should be done immediately or it's just going to get worse. I would also notify all developers of a hard-stop in submitting PRs until this is done.

Like I say, we've been holding off until towards the end of the end of the 0.10 maintenance period, but now we're coming up to 11.0 we're ready to merge it.

> Third, all existing PRs are going to need to be re-done. (This is a good reason why the PRs list shouldn't be allowed to get so long)

Possibly, yes, though git does cope surprisingly well with these types of moves.

> Will this make Avalonia.Base internal members visible to Avalonia.Controls? Access to things such as KnownColors isn't possible otherwise. Internals need to be made available IMO. I already have found a few places it's needed/useful and haven't done much in the code base.

`Avalonia.Base`'s internals are made visible to Avalonia.Controls, yes.

grokys commented on Directory.Build.props:
```diff
@@ -5,5 +5,6 @@
       <!-- https://github.com/dotnet/msbuild/issues/2661 -->
       <AddSyntheticProjectReferencesForSolutionDependencies>false</AddSyntheticProjectReferencesForSolutionDependencies>
       <MSBuildEnableWorkloadResolver>false</MSBuildEnableWorkloadResolver>
+      <RunApiCompat>False</RunApiCompat>
```
This PR currently disables ApiCompat. This is because even though git handles merges with file moves pretty well, it does __not__ handle changes to `ApiCompatBaseline.txt` well, and these files are going to be the main cause of merge conflicts after this change.

Should we leave this disabled until 11.0 (because breaking changes are expected at this point)?

robloo replied:
Well, the only issue I can think is its harder to make a breaking change summary for the release.

grokys committed:
Fix up merge errors in csprojs.

grokys committed:
Remove merged tests from nuke.

grokys committed:
Fix up namespace.

grokys committed:
Fix compile error.

grokys committed:
Fix failing tests.

Now that more unit tests are present in Avalonia.Base.UnitTests, general `AvaloniaObject` properties are getting registered. Ignore those.

MarchingCube reviewed (approved):
Let's get this merged!

grokys commented:
Anyone: if you're annoyed that your PR now has merge conflicts, ping me and I'll fix it ;)

Mrxx99 commented:
I think this does not have to be a breaking change when using the [TypeForwardedTo](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.typeforwardedtoattribute?view=net-6.0) Attribute, which is used very often in dotnet runtime itself for exactly that purpose
